File: java-se/pages/src/main/java/com/minte9/oop/polymorphism/di_note/DependencyInjection.java
package com.minte9.oop.polymorphism.di_note;

public class DependencyInjection {
    public static void main(String[] args) {
        Item csv = new Csv();
        Item xml = new Xml();

        FileOpener csvOpener = new FileOpener(csv);
        FileOpener xmlOpener = new FileOpener(xml);

        csvOpener.open();  // CSV opened
        xmlOpener.open();  // XML opened
       
    }
}

abstract class Item {
    public abstract void open();
}

class Csv extends Item {
    @Override
    public void open() {
        System.out.println("CSV opened");
    }
}

class Xml extends Item {
    @Override
    public void open() {
        System.out.println("XML opened");
    }
}

class FileOpener {
    private final Item item;  // dependency

    // Dependency Injection via constructor
    public FileOpener(Item item) {
        this.item = item;
    }

    public void open() {
        item.open();
    }
}

File: kotlin/bin/main/com/minte9/oop/polymorphism/DependencyInjection.kt
/**
 * Polymorphism + Dependency Injection in Kotlin
 *
 * FileOpener receive an Item in its constructor (DI).
 * At runtime, the same fileOpener.open() produces different
 * behavior depending on whether we injected Csv or Xml.
 */

package com.minte9.oop.polymorphism

fun main() {
    val csv = Csv()
    val xml = Xml()
    val csvOpener = FileOpener(csv)
    val xmlOpener = FileOpener(xml)
    csv.open()  // CSV opened
    xml.open()  // XML opened
}

abstract class Item {
    abstract fun open()
}

class Csv: Item() {
    override fun open() = println("CSV opened")
}

class Xml: Item() {
    override fun open() = println("XML opened")
}

class FileOpener (private val item: Item) {  // Dependency Injection
    fun open() = item.open()  // polymorphic call
}